go/core/action.go

```diff
@@ -28,6 +28,38 @@ import (
 	"github.com/invopop/jsonschema"
 )
 
+// Middleware is a function that takes in an action handler function and
+// returns a new handler function that might be changing input/output in
+// some way.
+//
+// Middleware functions can:
+//   - execute arbitrary code;
+//   - change the request and response;
+//   - terminate response by returning a response (or error);
+//   - call the next middleware function.
+type Middleware[I, O any] func(MiddlewareHandler[I, O]) MiddlewareHandler[I, O]
+
+type MiddlewareHandler[I, O any] func(ctx context.Context, input I) (O, error)
```

Comment:
Does this need to be exported?  If it does, it needs a doc comment.

Reply:
Added doc.